### ISSUE TITLE
MaterializedPostgreSQL: does not connect on failover

### ISSUE THREAD

**Issue by ben-pr-p** [ClickHouse/ClickHouse#28529]
> You have to provide the following information whenever possible.

**Describe what's wrong**

I have a PostgreSQL cluster with 2 databases, let's say A and B, hosted on Kubernetes. At any given point in time, there is a Kubernetes service that points to whichever database is the primary.

I set up my MaterializedPostgreSQL engine to point to that Kubernetes service, which when I set it up was pointing to A. At some point, several days later, a failover occurred and the primary had become B.

As a result, Clickhouse was no longer receiving things from the WAL logs, and the replication slot fell behind writes and started to pile up on disk, causing some problems on the primary.

I had to restart the Clickhouse pod in order to cause it to reconnect, which then properly reconnected to the primary.

**Does it reproduce on recent release?**

Version  21.8.4.51

**Enable crash reporting**

Unfortunately I lost logs when I killed the Clickhouse pod.

**How to reproduce**

Described above.

**Expected behavior**

The MaterializedPostgreSQL engine should detect when a failover has occurred (when it is connected to a read replica only) and reconnect.

**Additional context**

Tagging @kssenii . Very thankful for you work so far on this engine - it saved our team a lot of time setting up a custom sync, and is the reason we are choosing Clickhouse! Our sync time for a table of 30+ million rows in the same data center was under 10 minutes :), and our lag time has consistently been under 15 seconds even for inserts of 250k+ on PostgreSQL. Very impressive.

**Comment by kssenii** [comment]
> Very thankful for you work so far on this engine - it saved our team a lot of time setting up a custom sync, and is the reason we are choosing Clickhouse!

I'm glad its useful! :)

Ok,  documentation says that if there is a failover, then a new primary  will not be aware of any slots which were existing with old primary.
My first idea, before any further googling, is -- may be I can add a sql command which will restart replication for MaterializedPostgreSQL  database engine. But there is a huge drawback -- we will need to reload all the tables from scratch...
But it is the safest solution. I'm afraid there might be no safer solution because we cannot find out how much previous primary was lagging behind. I'll try to find a better solution, but can you say for now what do you think about the first one?

**Comment by bchrobot** [comment]
@kssenii we (I work with @ben-pr-p) are using [`postgres-operator`](https://github.com/zalando/postgres-operator) to manage our k8s clusters. The operator uses [`patroni`](https://github.com/zalando/patroni) under the hood.

It looks like we can address slot persistence via `patroni`'s `slot` setting to define permanent replication slots. From [`patroni`'s docs](https://patroni.readthedocs.io/en/latest/SETTINGS.html):

> These slots will be preserved during switchover/failover. The logical slots are copied from the primary to a standby with restart, and after that their position advanced every loop_wait seconds (if necessary). Copying logical slot files performed via libpq connection and using either rewind or superuser credentials (see postgresql.authentication section).

That would only work if we can manage the replication slots ourselves and provide those to MaterializedPostgreSQL rather than MaterializedPostgreSQL creating and managing slots itself.

It also comes with this warning:

> There is always a chance that the logical slot position on the replica is a bit behind the former primary, therefore application should be prepared that some messages could be received the second time after the failover. The easiest way of doing so - tracking `confirmed_flush_lsn`. 

Is my understanding correct that MaterializedPostgreSQL's `_version` virtual column addresses the above concern?

**Comment by kssenii** [comment]
> Is my understanding correct that MaterializedPostgreSQL's _version virtual column addresses the above concern?

Yes, correct.

> That would only work if we can manage the replication slots ourselves and provide those to MaterializedPostgreSQL

Ok, this would be a good solution and quick to implement.

**Comment by kssenii** [comment]
@ben-pr-p, hi, which clickhouse version do you use? I can backport there the changes which solve the issue.

**Comment by ben-pr-p** [comment]
> @ben-pr-p, hi, which clickhouse version do you use? I can backport there the changes which solve the issue.

Thanks so much! We're still getting started adopting it so not locked into anything - happy to just upgrade.

**Comment by kssenii** [comment]
ok, will backport to 21.8 then.

**Comment by bchrobot** [comment]
For anyone else using Spilo who stumbles across this GH issue, here's what we did.

Configure replication slot in `postgres-operator`/`spilo`:

```yaml
apiVersion: "acid.zalan.do/v1"
kind: postgresql
metadata:
  name: acid-demo-cluster
spec:
  numberOfInstances: 2
  postgresql:
    parameters:
      wal_level: logical
  patroni:
    slots:
      clickhouse_sync:
        type: logical
        database: demodb
        plugin: pgoutput
```

Wait for replication slot to be ready, then begin a transaction and export the transaction snapshot identifier:

```sql
begin;
select pg_export_snapshot();

-- 0000000A-0000023F-3
```

In ClickHouse, create database:

```sql
create database demodb
engine = MaterializedPostgreSQL('postgres1:5432', 'postgres_database', 'postgres_user', 'postgres_password')
settings
  materialized_postgresql_replication_slot = 'clickhouse_sync',
  materialized_postgresql_snapshot = '0000000A-0000023F-3',
  materialized_postgresql_tables_list = 'table1,table2,table3';
```

End the Postgres transaction once replication to ClickHouse DB is confirmed.

Verify that replication continues after failover:

```sh
kubectl exec acid-demo-cluster-0 -c postgres -- su postgres -c 'patronictl failover --candidate acid-demo-cluster-1 --force'
```

**Comment by bchrobot** [comment]
My understanding was that `materialized_postgresql_replication_slot` is only required for the initial table dump and not subsequent replication. It seems like this is not the case, however, and the MaterializedPostgreSQL database must be dropped and recreated after a ClickHouse restart.

We are seeing this error after CH restart:

```
2021.09.26 12:48:59.394661 [ 83 ] {} <Error> void DB::PostgreSQLReplicationHandler::waitConnectionAndStart(): std::exception. Code: 1001, type: pqxx::data_exception, e.what() = ERROR:  invalid snapshot identifier: "0000000A-0000023F-3"
```

The documentation suggests using `ALTER DATABASE ... MODIFY SETTING` but that throws a syntax error:

```
alter database demodb modify setting materialized_postgresql_snapshot = '0000000B-00000747-1';

Syntax error: failed at position 7 ('database'):

alter database demodb modify setting materialized_postgresql_snapshot = '0000000B-00000747-1';

Expected one of: PROFILE, ROLE, POLICY, USER, SETTINGS PROFILE, LIVE VIEW, ROW POLICY, QUOTA, TABLE
```

**Comment by kssenii** [comment]
> My understanding was that materialized_postgresql_replication_slot is only required for the initial table dump and not subsequent replication.

Yes, this is correct, forgot to modify some checks, sorry :( Will fix it right now and add more tests for these case.

> The documentation suggests using ALTER DATABASE ... MODIFY SETTING but that throws a syntax error:

becuase this feature was added one week ago and is not in any release yet.

**Comment by bchrobot** [comment]
> Yes, this is correct, forgot to modify some checks, sorry :( Will fix it right now and add more tests for usage of this setting.

Thank you! I'm not very familiar with the difference between `CREATE_REPLICATION_SLOT ... EXPORT_SNAPSHOT` and `pg_create_logical_replication_slot()` and thought that maybe the former created a persisted snapshot or something.

> becuase this feature was added one week ago and is not in any release yet.

Got it -- that makes sense :)

**Comment by kssenii** [comment]
> My understanding was that materialized_postgresql_replication_slot is only required for the initial table dump and not subsequent replication. It seems like this is not the case, however, and the MaterializedPostgreSQL database must be dropped and recreated after a ClickHouse restart.
We are seeing this error after CH restart

Hi, I finally found time to fix it and it turns out there is nothing to fix, because the behaviour is correct.

Replication slot must exist through the whole replication process, not only for initial tables dump. (Because its state must be updated after each batch poll of replication messages from wal and replication slot is a mechanism which makes it possible to do that, so replication will not work at all if the slot does not exist).

But I think it is not the reason why you got this exception, because looks like you did not delete this replication slot and it actually existed at the time as you would have noticed replication does not work. Probably at server restart something triggered deletion and then recreation of a replication slot (it is possible if there are some conflicts, but also I recently found one bug which triggers it, but it is only related if you use setting `materialzied_postgresql_tables_list`)

**Comment by bchrobot** [comment]
> Replication slot must exist through the whole replication process, not only for initial tables dump.

My apologies, @kssenii! I just realized my [comment](https://github.com/ClickHouse/ClickHouse/issues/28529#issuecomment-927307042) referenced the wrong configuration parameter. It should instead have said:

> My understanding was that `materialized_postgresql_snapshot` is only required for the initial table dump and not subsequent replication.

We don't delete the replication **slot**.

We are using `select pg_export_snapshot();` to generate the value for `materialized_postgresql_snapshot`. That snapshot identifier only exists for the lifetime of the PostgreSQL transaction, however, and we cannot leave the transaction open indefinitely. This seems to be referenced [here](https://github.com/kssenii/ClickHouse/blob/7b8101f289b33bce483a45bc6f85737225a644b2/src/Storages/PostgreSQL/PostgreSQLReplicationHandler.cpp#L620-L622).

If looks like `snapshot_name` is only used in [`loadFromSnapshot`](https://github.com/kssenii/ClickHouse/blob/7b8101f289b33bce483a45bc6f85737225a644b2/src/Storages/PostgreSQL/PostgreSQLReplicationHandler.cpp#L232) and [`reloadFromSnapshot`](https://github.com/kssenii/ClickHouse/blob/7b8101f289b33bce483a45bc6f85737225a644b2/src/Storages/PostgreSQL/PostgreSQLReplicationHandler.cpp#L605-L610). Maybe an issue with [`initial_sync` being called when it shouldn't](https://github.com/kssenii/ClickHouse/blob/7b8101f289b33bce483a45bc6f85737225a644b2/src/Storages/PostgreSQL/PostgreSQLReplicationHandler.cpp#L172-L179) for user-managed replication slot?

**Comment by kssenii** [comment]
Yes, I agree.
[This place](https://github.com/kssenii/ClickHouse/blob/7b8101f289b33bce483a45bc6f85737225a644b2/src/Storages/PostgreSQL/PostgreSQLReplicationHandler.cpp#L172-L179) you found is indeed the place where it ended up, though shouldn't have.
`new_publication` case if more precise -- it means list of tables in `publication` differs from the list of tables in setting `materialized_postgresql_tables_list`.  And I found a bug in this place just yesterday -- comparison was done incorrectly and this triggered `new_publication = true` and ending up in [this branch](https://github.com/kssenii/ClickHouse/blob/7b8101f289b33bce483a45bc6f85737225a644b2/src/Storages/PostgreSQL/PostgreSQLReplicationHandler.cpp#L172-L179).
So am I correct and you were indeed using setting `materialized_postgresql_tables_list`?
I think I'll remove this `new_publication` case completely to avoid such unwanted behavour, and also will fix the bug I found which triggered your error. Hope it will be alright then.
